### DISCUSSION TITLE
Should I go with InvokeAsync as much as possible?

### DISCUSSION THREAD

**Discussion by dziedrius** [microsoft/semantic-kernel#6538] [Q&A] [answered]
Hi, still very new to this and I think some of the abstraction is puzzling me.
Maybe I overlooked, but I could not find explanation when to use which abstraction and to me it seems that sometimes there's an abstraction leak.

For example, I can execute chat completion in several ways:
```
kernel.InvokePromptAsync(... // directly
kernel.InvokeAsync(... // through kernel function KernelFunctionFactory.CreateFromPrompt
kernel.GetRequiredService<IChatCompletionService>().GetChatMessageContentsAsync(...
```
So initially, I was trying to use kernel.InvokeAsync, as it seemed like a most generic way and at least in some tutorials that was the way, as it supports plugin conventions. 
Then I spent quite a while searching how to pass chat history into it - may be some convention how to name some kernel argument or something that is stored and expected in external context, etc. Finally going through source code I found, that it won't accept chat history, as deep inside it does this:
```
if (result.AIService is IChatCompletionService chatCompletion)
{
     asyncReference = chatCompletion.GetStreamingChatMessageContentsAsync(result.RenderedPrompt, result.ExecutionSettings, kernel, cancellationToken);
}
```
Hence it seems that if I need to attach chat history, I need to go with `kernel.GetRequiredService<IChatCompletionService>().GetChatMessageContentsAsync(...`, but I was wondering if there is some kind of guidance when to use what.

**Comment by madsbolaris** [comment, marked as answer]
Yes, as of _today_, you'll need to use the underlying service to do anything with the chat history object. This is currently flagged as a gap that we'd like to solve this summer. Our goal is to provide helpers within prompt templates that allow you to write prompts _with_ chat history objects. Once that's available, you'll be able to use `InvokePromptAsync`. We should have an initial proposal out in a few weeks.
